ProjFS.Mac/PrjFSKext/PrjFSKext/KextLog.hpp

```diff
@@ -42,6 +44,53 @@ template <typename... args>
 #define KextLog_FileInfo(vnode, format, ...)  ({ _os_log_verify_format_str(format, ##__VA_ARGS__); KextLogFile_Printf(KEXTLOG_INFO, vnode, format " (vnode path: '%s')", ##__VA_ARGS__); })
 #define KextLog_FileNote(vnode, format, ...)  ({ _os_log_verify_format_str(format, ##__VA_ARGS__); KextLogFile_Printf(KEXTLOG_NOTE, vnode, format " (vnode path: '%s')", ##__VA_ARGS__); })
 
+template <typename... args>
+    void KextLog_PrintfVnodePathAndProperties(KextLog_Level loglevel, struct vnode* vnode, const char* fmt, args... a)
+    {
+        char vnodePath[PrjFSMaxPath] = "";
+        int vnodePathLength = PrjFSMaxPath;
+        vn_getpath(vnode, vnodePath, &vnodePathLength);
+        
+        const char* name = vnode_getname(vnode);
+        mount_t mount = vnode_mount(vnode);
+        vfsstatfs* vfsStat = mount != nullptr ? vfs_statfs(mount) : nullptr;
+        
+        KextLog_Printf(loglevel, fmt, a..., vnodePath, name ?: "[NULL]", vnode_vtype(vnode), vnode_isrecycled(vnode) ? "yes" : "no", vfsStat ? vfsStat->f_mntonname : "[NULL]");
+ 
+        if (name != nullptr)
+        {
+            vnode_putname(name);
+        }
+    }
+
+template <typename... args>
+    void KextLog_PrintfVnodeProperties(KextLog_Level loglevel, struct vnode* vnode, const char* fmt, args... a)
+    {
+        const char* name = vnode_getname(vnode);
+        mount_t mount = vnode_mount(vnode);
+        vfsstatfs* vfsStat = mount != nullptr ? vfs_statfs(mount) : nullptr;
+        
+        KextLog_Printf(loglevel, fmt, a..., name ?: "[NULL]", vnode_vtype(vnode), vnode_isrecycled(vnode) ? "yes" : "no", vfsStat ? vfsStat->f_mntonname : "[NULL]");
```

Comment:
Is there any clean way to move the format strings out of the macros and into `KextLog_PrintfVnodePathAndProperties` and `KextLog_PrintfVnodeProperties`?

I like the addition of the helpers, but the one thing that threw me off a bit was that the `KextLog_PrintfVnodePathAndProperties` and `KextLog_PrintfVnodeProperties` *must* be called with a particular `fmt`.

Reply:
> Is there any clean way to move the format strings out of the macros and into `KextLog_PrintfVnodePathAndProperties` and `KextLog_PrintfVnodeProperties`?

Essentially, no, at least not without runtime downsides. Let's look at the macro definition:
```
#define KextLog_ErrorVnodeProperties(vnode, format, ...) \
    ({ \
        _os_log_verify_format_str(format, ##__VA_ARGS__); \
        KextLog_PrintfVnodeProperties(KEXTLOG_ERROR, vnode, format " (vnode name: '%s', type: %d, recycling: %s, mount point mounted at path '%s')", ##__VA_ARGS__); \
    })
```

Note that the `format` macro parameter is placed directly next to the extra format string literal with only whitespace as separation:
```
format " (vnode name: '%s', type: %d, recycling: %s, mount point mounted at path '%s')"
```
When expanded, this simply places 2 string literals next to one another, which the compiler concatenates into a single string literal - this means it wouldn't won't work if you passed a symbolic format string variable or constant. (except of course a macro that itself expands to a string literal)

This literal concatenation of course doesn't work on function parameters, so it can't go inside the helper function. As far as I'm aware, you can't sensibly do this even with C++14's `constexpr` functions. There are some [awful hacks with templates](https://akrzemi1.wordpress.com/2017/06/28/compile-time-string-concatenation/), but for this use case it would make things much worse, not better.

> I like the addition of the helpers, but the one thing that threw me off a bit was that the `KextLog_PrintfVnodePathAndProperties` and `KextLog_PrintfVnodeProperties` _must_ be called with a particular `fmt`.

The other possible solution to that is of course to not use a helper function in the first place, and just put everything in the macro. I very much wanted to avoid that especially for the cases where we call `vn_getpath` with a stack-allocated string buffer. If we created one of those in an expanded macro, and we used that macro multiple times in a function, the compiler could elect to statically allocate space for those temporary buffers in the stack frame. We have a very small stack in kernel threads (a little less than 16KiB last time I checked), and the path string buffer is very large (1KiB) so allocating a bunch of them could
easily blow the stack, especially with nested function calls. Putting the buffer in a near-leaf helper function minimises that risk.

So I agree, it's not pretty, but it's the least bad solution I could think of.

Reply:
I've just pushed a new version of this PR which adds this rationale as a comment for the `KextLogFile_Printf()` function, where the same reasoning applies.